pkg/controller/cloudfront/cloudfrontoriginaccessidentity/setup.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloudfrontoriginaccessidentity
+
+import (
+	"context"
+	"time"
+
+	svcsdk "github.com/aws/aws-sdk-go/service/cloudfront"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/google/go-cmp/cmp"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/cloudfront/v1alpha1"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+// SetupCloudFrontOriginAccessIDentity adds a controller that reconciles CloudFrontOriginAccessIDentity .
+func SetupCloudFrontOriginAccessIDentity(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(svcapitypes.CloudFrontOriginAccessIDentityGroupKind)
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewController(rl),
+		}).
+		For(&svcapitypes.CloudFrontOriginAccessIDentity{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.CloudFrontOriginAccessIDentityGroupVersionKind),
+			managed.WithExternalConnecter(&connector{
+				kube: mgr.GetClient(),
+				opts: []option{
+					func(e *external) {
+						e.preObserve = preObserve
+						e.postObserve = postObserve
+						e.preCreate = preCreate
+						e.postCreate = postCreate
+						e.lateInitialize = lateInitialize
+						e.preUpdate = preUpdate
+						e.isUpToDate = isUpToDate
+						e.preDelete = preDelete
+					},
+				},
+			}),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+func preCreate(_ context.Context, cr *svcapitypes.CloudFrontOriginAccessIDentity, cdi *svcsdk.CreateCloudFrontOriginAccessIdentityInput) error {
+	cdi.CloudFrontOriginAccessIdentityConfig.CallerReference = awsclients.String(string(cr.UID))
```

Comment:
What is CallerReference used for? UID will change if the managed resource was deleted with `deletionPolicy: Orphan` then recreated to adopt the existing external resource by setting an `external-name` annotation. Would that matter?

Reply:
This is the same pattern used with CloudFrontDistribution: https://github.com/crossplane/provider-aws/blob/master/pkg/controller/cloudfront/distribution/setup.go#L83

According to the [API](https://docs.aws.amazon.com/cloudfront/latest/APIReference/API_CreateCloudFrontOriginAccessIdentity.html#cloudfront-CreateCloudFrontOriginAccessIdentity-request-CallerReference) it is "A unique value (for example, a date-time stamp) that ensures that the request can't be replayed." 

The caller reference is not present in the AWS UI, but the AWS cli allows setting it. It is a required setting, AWS will return an error if not present. 

We could remove CallerReference from the ignored_field path and allow users to set it manually instead of having the provider generate a random value.

Reply:
Got it. Sounds like UID should be appropriate here. Thank you!